server/schedulers/shuffle_region_config.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/pingcap/pd/v4/pkg/apiutil"
+	"github.com/pingcap/pd/v4/pkg/slice"
+	"github.com/pingcap/pd/v4/server/core"
+	"github.com/pingcap/pd/v4/server/schedule"
+	"github.com/pingcap/pd/v4/server/schedule/placement"
+	"github.com/unrolled/render"
+)
+
+const (
+	roleLeader   = string(placement.Leader)
+	roleFollower = string(placement.Follower)
+	roleLearner  = string(placement.Learner)
+)
+
+var allRoles = []string{roleLeader, roleFollower, roleLearner}
+
+type shuffleRegionSchedulerConfig struct {
+	sync.RWMutex
+	storage *core.Storage
+
+	Ranges []core.KeyRange `json:"ranges"`
+	Roles  []string        `json:"roles"` // can include `leader`, `follower`, `learner`.
+}
+
+func (conf *shuffleRegionSchedulerConfig) EncodeConfig() ([]byte, error) {
+	conf.RLock()
+	defer conf.RUnlock()
+	return schedule.EncodeConfig(conf)
+}
+
+func (conf *shuffleRegionSchedulerConfig) GetRoles() []string {
+	conf.RLock()
+	defer conf.RUnlock()
+	return conf.Roles
+}
+
+func (conf *shuffleRegionSchedulerConfig) GetRanges() []core.KeyRange {
+	conf.RLock()
+	defer conf.RUnlock()
+	return conf.Ranges
+}
+
+func (conf *shuffleRegionSchedulerConfig) IsRoleAllow(role string) bool {
+	conf.RLock()
+	defer conf.RUnlock()
+	return slice.AnyOf(conf.Roles, func(i int) bool { return conf.Roles[i] == role })
+}
+
+func (conf *shuffleRegionSchedulerConfig) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	router := mux.NewRouter()
+	router.HandleFunc("/roles", conf.handleGetRoles).Methods("GET")
```

Comment:
Is this the root path of  PD's API? Maybe add the PD API prefix is better?

Reply:
no, it is a sub router to be registered under `scheduler_config/shuffle-region-scheduler/`.